### PR TITLE
Fix HasOneThrough / HasManyThrough relationships

### LINKED ISSUE #1086: Count parameter not working with Has-Many-Through relation
### Winter CMS Build

1.2.5

### PHP Version

8.1

### Database engine

MySQL/MariaDB

### Plugins installed

_No response_

### Issue description

Hi,

I have the following scenario: a `Project` has **only one** `Board` which has many `Tasks`. I would like to get the count of all tasks belonging to a project. It's like WinterCMS is ignoring the `count` parameter because what I get is an array of `Task` models.

`Project` model:
```php
public $hasManyThrough = [
	'tasks_count' => [
		Task::class,
		'through' => Board::class,
		'key' => 'project_id',
		'throughKey' => 'board_id',
		'scope' => 'completed',
		'count' => true,
	]
];
```

The scope `completed` is ignored too.

`Task` model:
```php
public function scopeCompleted($query)
{
	return $query->where('status', TaskStatus::Completed);
}
```

### Steps to replicate

Set up a `Has-Many-Through` relation and try to get the `count`.

### Workaround

I had to set up the `Has-Many-Through` relation without `count` and `scope` parameters and then use `$appends` to insert the count value into the response.

`Project` model:
```php
public $appends = [
	'tasks_completed_count'
];

public $hasManyThrough = [
	'tasks' => [
		Task::class,
		'through' => Board::class,
		'key' => 'project_id',
		'throughKey' => 'board_id'
	]
];

public function getTasksCompletedCountAttribute()
{
	return $this->tasks()->completed()->count();
}
```

### PR DESCRIPTION
**Pull request by mjauvin**
Fixes https://github.com/wintercms/winter/issues/1086

Take `count` & `scope` relation configs into account.

Also pass the relation name to the constructor.

### CODE REVIEW

**Comment by mjauvin** [comment]
@goldmont, can you test if this works for you ?

Keep in mind your scope will need to fully qualify its field names if they are not unique since this created a join query between multiple tables.

**Comment by mjauvin** [comment]
@bennothommo phpstan is complaining about a method that does not exist even though it does.

Not sure about the error on Windows, though...

**Comment by goldmont** [comment]
> @goldmont, can you test if this works for you ?
> 
> Keep in mind your scope will need to fully qualify its field names if they are not unique since this created a join query between multiple tables.

Hi,

It works now!

**Comment by goldmont** [comment]
@mjauvin When will it be released? Thanks.

**Comment by mjauvin** [comment]
> @mjauvin When will it be released? Thanks.

Not sure. It should be easy to get merged to dev-develop, released is another story as there are no release planned short term.

**Comment by goldmont** [comment]
> > @mjauvin When will it be released? Thanks.
> 
> Not sure. It should be easy to get merged to dev-develop, released is another story as there are no release planned short term.

Ok, understanding 👍🏻

**Comment by mjauvin** [comment]
@bennothommo did I ever tell you how much I hate phpstan ? :stuck_out_tongue: 

It's running fine locally now, but barfs in github workflow (see screenshot below):

image

**Comment by bennothommo** [comment]
@mjauvin which PHP version are you running it on? The GitHub Actions one runs it on PHP 8.2

**Comment by bennothommo** [comment]
Absolutely no idea why ignoring the `KeyGenerateCommand` file in Foundation made it work, but it did, so go team!

**Comment by mjauvin** [comment]
> @mjauvin which PHP version are you running it on? The GitHub Actions one runs it on PHP 8.2

Most likely 8.3 but I'll need to check tomorrow

**Comment by mjauvin** [comment]
> @mjauvin which PHP version are you running it on? The GitHub Actions one runs it on PHP 8.2

As I thought, this was using php 8.3:

```bash
[r4l@ahs5] ~/winter/storm $ php -v
PHP 8.3.4 (cli) (built: Mar 15 2024 14:04:12) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.3.4, Copyright (c) Zend Technologies
    with Zend OPcache v8.3.4, Copyright (c), by Zend Technologies
```

File: src/Database/Relations/Concerns/DefinedConstraints.php [hunk 1]
```diff
@@ -1,6 +1,7 @@
 <?php namespace Winter\Storm\Database\Relations\Concerns;
 
 use Illuminate\Database\Eloquent\Relations\BelongsToMany as BelongsToManyBase;
+use Illuminate\Database\Eloquent\Relations\HasManyThrough as HasManyThroughBase;
```
Comment (LukeTowers) [inline src/Database/Relations/Concerns/DefinedConstraints.php]:
```suggestion
```

**Pull request merged**